agent/pool/pool.go

```diff
@@ -55,6 +55,47 @@ type Conn struct {
 	clientLock sync.Mutex
 }
 
+// TimeoutConn wraps net.Conn with a default timeout.
+// readTimeout and writeTimeout apply to the subsequent operation.
+type TimeoutConn struct {
+	net.Conn
+	DefaultTimeout time.Duration
+	ReadTimeout    time.Duration
+	WriteTimeout   time.Duration
+}
+
+func (c *TimeoutConn) Read(b []byte) (int, error) {
+	timeout := c.ReadTimeout
+	c.ReadTimeout = 0
+	if timeout == 0 {
+		timeout = c.DefaultTimeout
+	}
+	deadline := time.Time{}
+	if timeout > 0 {
+		deadline = time.Now().Add(timeout)
+	}
+	if err := c.Conn.SetReadDeadline(deadline); err != nil {
+		return 0, err
+	}
+	return c.Conn.Read(b)
+}
```

Comment:
Probably out of scope for this PR but wondering if there's any room for optimization here and in `Write` by reducing allocations.